src/main/scala/org/broadinstitute/dsde/workbench/avram/api/AvramServlet.scala

```diff
@@ -0,0 +1,67 @@
+package org.broadinstitute.dsde.workbench.avram.api
+
+
+import scala.concurrent.duration.Duration
+import javax.servlet.http.{HttpServletRequest, HttpServletResponse}
+
+import org.broadinstitute.dsde.workbench.avram.Avram
+import org.broadinstitute.dsde.workbench.avram.dataaccess.SamUserInfoResponse
+import org.broadinstitute.dsde.workbench.avram.model.AvramException
+import org.broadinstitute.dsde.workbench.avram.util.ErrorResponse
+
+import scala.concurrent.Await
+import scala.util.{Failure, Success}
+//import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future
+
+trait AvramServlet {
+
+  implicit val executionContext = scala.concurrent.ExecutionContext.Implicits.global
+
+  private val samDao = Avram.samDao
+  private val bearerPattern = """(?i)bearer (.*)""".r
+
+  def registerOnComplete[T]
+  (future: Future[T], request: HttpServletRequest, response: HttpServletResponse)
+  (f: T => Unit): Future[T] = {
+    future onComplete {
+      case Success(result) => f(result)
+      case Failure(e: AvramException) => {
+        response.setStatus(e.status)
+        response.getWriter.write(e.message)
+      }
+      case Failure(e) => response.getWriter.write(e.getMessage)
+    }
+    future
+  }
+
+
+  def handleAuthenticatedRequest[T]
+  (request: HttpServletRequest, response: HttpServletResponse)
+  (f: SamUserInfoResponse => Future[Either[AvramException, T]]): T = {
+    val thing = for {
+      userInfo <- extractUserInfo(request)
+      result <- f(userInfo)
+    } yield userInfo
+    unsafeRun(Await.result(f(thing), Duration.apply(30, "second")))
+  }
```

Comment:
Changed this to take in a Future but I'm having a little trouble wrapping my head around executing the Future and the Either[Exception, T] format and how best to structure it so that it does what I want. Possibly better as a face to face conversation

Reply:
`unsafeRun` should wrap anything that causes an effect (including fetching user info from Sam), ultimately leading to an `Either[AvramException, T]`. `f` is a function from `SamUserInfoResponse` to `Future[Either[AvramException, T]]`. All that's needed is to get the user info, invoke the function that returns a future, and await the result and we'll end up with an `Either[AvramException, T]`. So:
```
    unsafeRun {
      for {
        userInfo <- extractUserInfo(request)
        result <- Await.result(f(userInfo), Duration.apply(30, "second"))
      } yield result
    }
```
Though now that I look at it, `unsafeRun` isn't a good name for this function when not using something like cats `IO`. It's really just `eitherToResultOrThrow`. I should have renamed it or inlined the code.

It was easier for me to give an initial written reply but I'm happy to talk face-to-face too if that helps.

Reply:
i'd like to understand this more. especially if we really need the Await and the name of the method. let's talk faces